File: androidApp/src/main/java/me/nathanfallet/bdeensisa/features/shop/ShopViewModel.kt
package me.nathanfallet.bdeensisa.features.shop

import android.app.Application
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.viewModelScope
import com.google.firebase.analytics.FirebaseAnalytics
import com.google.firebase.analytics.ktx.analytics
import com.google.firebase.analytics.ktx.logEvent
import com.google.firebase.ktx.Firebase
import kotlinx.coroutines.launch
import me.nathanfallet.bdeensisa.database.DatabaseDriverFactory
import me.nathanfallet.bdeensisa.extensions.SharedCacheService
import me.nathanfallet.bdeensisa.models.CotisantConfiguration
import me.nathanfallet.bdeensisa.models.TicketConfiguration

class ShopViewModel(application: Application) : AndroidViewModel(application) {

    // Properties

    private val cotisantConfigurations = MutableLiveData<List<CotisantConfiguration>>()
    private val ticketConfigurations = MutableLiveData<List<TicketConfiguration>>()

    // Getters

    fun getCotisantConfigurations(): LiveData<List<CotisantConfiguration>> {
        return cotisantConfigurations
    }

    fun getTicketConfigurations(): LiveData<List<TicketConfiguration>> {
        return ticketConfigurations
    }

    // Methods

    init {
        Firebase.analytics.logEvent(FirebaseAnalytics.Event.SCREEN_VIEW) {
            param(FirebaseAnalytics.Param.SCREEN_NAME, "shop")
            param(FirebaseAnalytics.Param.SCREEN_CLASS, "ShopView")
        }

        fetchData()
    }

    fun fetchData() {
        viewModelScope.launch {
            try {
                SharedCacheService.getInstance(DatabaseDriverFactory(getApplication())).apiService()
                    .getCotisantConfigurations().let {
                        cotisantConfigurations.value = it
                    }
            } catch (e: Exception) {
                e.printStackTrace()
            }
        }
        viewModelScope.launch {
            try {
                SharedCacheService.getInstance(DatabaseDriverFactory(getApplication())).apiService()
                    .getTicketConfigurations().let {
                        ticketConfigurations.value = it
                    }
            } catch (e: Exception) {
                e.printStackTrace()
            }
        }
    }

}

File: iosApp/iosApp/Features/Shop/ShopViewModel.swift
//
//  ShopViewModel.swift
//  BDE
//
//  Created by Nathan Fallet on 22/03/2023.
//  Copyright © 2023 orgName. All rights reserved.
//

import Foundation
import shared

class ShopViewModel: ObservableObject {
    
    @Published var cotisantConfigurations = [CotisantConfiguration]()
    @Published var ticketConfigurations = [TicketConfiguration]()
    
    func onAppear() {
        AnalyticsService.shared.log(.screenView(screenName: "shop", screenClass: "ShopView"))
        
        fetchData()
    }
    
    func fetchData() {
        Task {
            let cotisantConfigurations = try await CacheService.shared.apiService().getCotisantConfigurations()
            DispatchQueue.main.async {
                self.cotisantConfigurations = cotisantConfigurations
            }
        }
        Task {
            let ticketConfigurations = try await CacheService.shared.apiService().getTicketConfigurations()
            DispatchQueue.main.async {
                self.ticketConfigurations = ticketConfigurations
            }
        }
    }
    
}
